File: sorting/test_sorting.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include "sorting.h"


void test_sorting(void (*func)(int *, int), char *func_name, int size) {
    int s[size];
    srand(time(NULL));
    printf("Test %s\n", func_name);
    for (int i = 0; i < size; i++) {
    /* Note this is not uniformly distributed */
        s[i] = rand() % size;
        printf(" %d", s[i]);
    }
    printf("\n");
    func(s, size);
    for (int i = 0; i < size; i++) {
        printf(" %d", s[i]);
    }
    printf("\n");
}


int main() {
    const int num = 7;
    void (*func[])(int *, int) = {&bubble_sort, &insertion_sort,
                                     &selection_sort, &shell_sort,
                                     &merge_sort, &quick_sort_kr,
                                     &quick_sort_sedgewick};
    char *func_name[] = {"bubble_sort", "insertion_sort",
                            "selection_sort", "shell_sort",
                            "merge_sort", "quick_sort_kr",
                            "quick_sort_sedgwick"};
    int size = 20;
    for (int i = 0; i < num; i++) {
        test_sorting(func[i], func_name[i], size);
    }
    
    return 0;
}

File: sorting/test_sorting.py
#/usr/bin/env python3

from elementarysort import bubble_sort, selection_sort, insertion_sort
from quicksort import quick_sort_kr, quick_sort_sedgewick 
from mergesort import merge_sort
from shellsort import shell_sort_kr, shell_sort_shell, shell_sort_sedgewick

def test_sorting(func, size, duplicate=False):
    from random import shuffle, randint
    if duplicate:
        s = list(map(lambda x: randint(0, size-1), range(size)))
    else:
        s = list(range(size))
    shuffle(s)
    print('Testing %s'%func.__name__)
    print('----------------------')
    print(s)
    try:
        func(s)
    except TypeError: 
        func(s, 0, size-1)
    print(s)
    print('----------------------')
    
if __name__ == '__main__':
    funcs = [bubble_sort, selection_sort, insertion_sort, merge_sort,
             quick_sort_kr, quick_sort_sedgewick, shell_sort_kr,
             shell_sort_shell, shell_sort_sedgewick]
    size = 20
    duplicate = True
    for func in funcs:
        test_sorting(func, size, duplicate)
